File: 1673-find-the-most-competitive-subsequence/1673-find-the-most-competitive-subsequence.java
class Solution {
    public int[] mostCompetitive(int[] nums, int k) {
        int[] ans=new int[k];
        int rem=nums.length-k;
        int top=-1;
        for(int a:nums){
            while(top>=0&& rem>0&& a<ans[top]){
                top--;
                rem--;
            }
            if(top<k-1){
                ans[++top]=a;
            }
            else{
                rem--;
            }
        }
        return ans;
    }
}

File: 1673-find-the-most-competitive-subsequence/1673-find-the-most-competitive-subsequence.cpp
class Solution {
public:
    vector<int> mostCompetitive(vector<int>& nums, int k) {
        vector<int> ans;
        int rem=nums.size()-k;
        for(auto a:nums){
            while(!ans.empty()&& rem>0&& a<ans.back()){
                ans.pop_back();
                rem--;
            }
            ans.push_back(a);
        }
        while(ans.size()>k){
            ans.pop_back();
        }
        return ans;
    }
};